File: acp-sdk-python/acp/transport.py
from __future__ import annotations

from typing import Any

import requests


class TransportError(RuntimeError):
    pass


class HTTPTransport:
    def __init__(self, *, timeout_seconds: int = 10) -> None:
        self.timeout_seconds = timeout_seconds

    def post_json(self, url: str, body: dict[str, Any]) -> requests.Response:
        try:
            return requests.post(url, json=body, timeout=self.timeout_seconds)
        except requests.RequestException as exc:
            raise TransportError(f"HTTP POST failed for {url}: {exc}") from exc

    def get_json(self, url: str, params: dict[str, str] | None = None) -> dict[str, Any]:
        try:
            response = requests.get(url, params=params, timeout=self.timeout_seconds)
        except requests.RequestException as exc:
            raise TransportError(f"HTTP GET failed for {url}: {exc}") from exc
        if response.status_code != 200:
            raise TransportError(f"HTTP GET {url} returned {response.status_code}")
        try:
            return response.json()
        except ValueError as exc:
            raise TransportError(f"HTTP GET {url} did not return JSON") from exc

File: acp-sdk-typescript/src/transport.ts
import { AcpMessage, messageToMap } from "./messages.js";
import { HttpSecurityPolicy, buildFetchOptions, validateHttpUrl } from "./httpSecurity.js";
import { JsonMap, parseJsonMap } from "./jsonSupport.js";
import { transportError } from "./errors.js";

export interface TransportResponse {
  status_code: number;
  body?: JsonMap;
  raw_body: string;
}

export class TransportClient {
  private readonly fetchOptions: ReturnType<typeof buildFetchOptions>;

  public constructor(
    private readonly timeoutSeconds: number,
    private readonly policy: HttpSecurityPolicy
  ) {
    this.fetchOptions = buildFetchOptions(policy);
  }

  public async postJson(url: string, body: JsonMap): Promise<TransportResponse> {
    validateHttpUrl(url, this.policy.allow_insecure_http, this.policy.mtls_enabled, "HTTP transport request");
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), Math.max(1, this.timeoutSeconds) * 1000);
    try {
      const response = await fetch(url, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(body),
        signal: controller.signal,
        ...this.fetchOptions
      } as RequestInit);
      const rawBody = await response.text();
      let parsedBody: JsonMap | undefined;
      try {
        parsedBody = parseJsonMap(rawBody);
      } catch {
        parsedBody = undefined;
      }
      return {
        status_code: response.status,
        raw_body: rawBody,
        body: parsedBody
      };
    } catch (error) {
      throw transportError(`HTTP request failed: ${String(error)}`);
    } finally {
      clearTimeout(timeout);
    }
  }

  public async sendToRelay(relayUrl: string, message: AcpMessage): Promise<JsonMap> {
    const relayEndpoint = relayUrl.endsWith("/") ? `${relayUrl}messages` : `${relayUrl}/messages`;
    const response = await this.postJson(relayEndpoint, messageToMap(message));
    if (response.status_code !== 200) {
      throw transportError(
        `Relay returned HTTP ${response.status_code} for message ${message.envelope.message_id}`
      );
    }
    if (!response.body) {
      throw transportError("Relay returned non-JSON response");
    }
    return response.body;
  }
}
